src/Analyzer/Passes/RewriteSumFunctionWithSumAndCountPass.cpp

```diff
@@ -0,0 +1,135 @@
+#include <Analyzer/Passes/RewriteSumFunctionWithSumAndCountPass.h>
+
+#include <AggregateFunctions/AggregateFunctionFactory.h>
+#include <Analyzer/InDepthQueryTreeVisitor.h>
+#include <Analyzer/ColumnNode.h>
+#include <Analyzer/ConstantNode.h>
+#include <Analyzer/FunctionNode.h>
+#include <Functions/FunctionFactory.h>
+
+namespace DB
+{
+
+/** Rewrite the following Analyzer to break the function `sum(column +/- literal)` into two individual functions
+ * `sum(column)` and `literal * count(column)`.
+ * sum(column + literal) -> sum(column) + literal * count(column)
+ * sum(literal + column) -> literal * count(column) + sum(column)
+ * sum(column - literal) -> sum(column) - literal * count(column)
+ * sum(literal - column) -> literal * count(column) - sum(column)
+ */
+
+namespace
+{
+
+class RewriteSumFunctionWithSumAndCountVisitor : public InDepthQueryTreeVisitorWithContext<RewriteSumFunctionWithSumAndCountVisitor>
+{
+public:
+    using Base = InDepthQueryTreeVisitorWithContext<RewriteSumFunctionWithSumAndCountVisitor>;
+    using Base::Base;
+
+    void enterImpl(QueryTreeNodePtr & node)
+    {
+        static const std::unordered_set<String> func_supported = {
+            "plus",
+            "minus"
+        };
+
+        auto * function_node = node->as<FunctionNode>();
+        if (!function_node || Poco::toLower(function_node->getFunctionName()) != "sum")
+            return;
+
+        const auto & function_nodes = function_node->getArguments().getNodes();
+        if (function_nodes.size() != 1)
+            return;
+
+        const auto * func_plus_minus_node = function_nodes[0]->as<FunctionNode>();
+        if (!func_plus_minus_node || !func_supported.contains(Poco::toLower(func_plus_minus_node->getFunctionName())))
+            return;
+
+        const auto & func_plus_minus_nodes = func_plus_minus_node->getArguments().getNodes();
+        if (func_plus_minus_nodes.size() != 2)
+            return;
+
+        size_t column_id = func_plus_minus_nodes.size();
+        for (size_t i = 0; i < func_plus_minus_nodes.size(); i++)
+        {
+            if (const auto * column_node = func_plus_minus_nodes[i]->as<ColumnNode>())
+                column_id = i;
+        }
+        if (column_id == func_plus_minus_nodes.size())
+            return;
+
+        size_t literal_id = 1 - column_id;
+        const auto * literal = func_plus_minus_nodes[literal_id]->as<ConstantNode>();
+        if (!literal)
+            return;
+
+        const auto literal_type = literal->getResultType();
+        if (!literal_type || !WhichDataType(literal_type).isNumber())
+            return;
+
+        const auto * column_node = func_plus_minus_nodes[column_id]->as<ColumnNode>();
+        if (!column_node)
```

Comment:
How is this possible?

Reply:
This part might still be needed in the latest code. Could you help to review the latest code?